### PR TITLE
OCPBUGS-32930: Allow specifying resource requests per cluster size

### PR DESCRIPTION
**Pull request by csrwng**
**What this PR does / why we need it**:
Allows memory and CPU requests to be specified by deployment/container per size.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #[OCPBUGS-32930](https://issues.redhat.com/browse/OCPBUGS-32930)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go [hunk 1]
```diff
@@ -617,6 +617,11 @@ func (r *DedicatedServingComponentSchedulerAndSizer) updateHostedCluster(ctx con
 		hc.Annotations[hyperv1.APICriticalPriorityClass] = *sizeConfig.Effects.APICriticalPriorityClassName
 	}
 
+	resourceRequestAnnotations := resourceRequestsToOverrideAnnotations(sizeConfig.Effects.ResourceRequests)
+	for k, v := range resourceRequestAnnotations {
+		hc.Annotations[k] = v
+	}
```
Comment (muraee) [inline hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go:623]: for my understanding, why are we using annotations? can't the CPO read the clustersizing CR directly?

File: api/scheduling/v1alpha1/clustersizingconfiguration_types.go [hunk 2]
```diff
@@ -177,6 +181,24 @@ type TransitionDelayConfiguration struct {
...
+	// DeploymentName is the name of the deployment to which the resource request applies.
+	DeploymentName string `json:"deploymentName"`
+
+	// +kubebuilder:validation:Required
+	// ContainerName is the name of the container to which the resource request applies.
+	ContainerName string `json:"containerName"`
+
+	// +kubebuilder:validation:Optional
+	// Memory is the amount of memory to request for the container.
+	Memory *resource.Quantity `json:"memory,omitempty"`
```
Comment (muraee) [inline api/scheduling/v1alpha1/clustersizingconfiguration_types.go:195]: This field can also be used to configure KAS, right? can we remove/deprecate `KASMemoryRequest` field, then? https://github.com/openshift/hypershift/blob/a72e5814d121aeadc03d573b3c714c6aac7d84e8/api/scheduling/v1alpha1/clustersizingconfiguration_types.go#L106

Comment (csrwng) [reply api/scheduling/v1alpha1/clustersizingconfiguration_types.go:195]: Yup, makes sense. AFAIK we're not using the KASMemoryRequest field, so I will just remove it. Thx!

File: hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go (hunk 1 above)
Comment (muraee) [reply hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go:623]: thanks.

File: api/scheduling/v1alpha1/clustersizingconfiguration_types.go (hunk 2 above)
Comment (csrwng) [reply api/scheduling/v1alpha1/clustersizingconfiguration_types.go:191]: Fixed

**Comment by csrwng** [comment]
/override "Red Hat Konflux / hypershift-operator-main-on-pull-request"

**Pull request merged**